synapse/crypto/context_factory.py

```diff
@@ -127,9 +127,32 @@ class ClientTLSOptionsFactory(object):
     to remote servers for federation."""
 
     def __init__(self, config):
-        # We don't use config options yet
-        self._options = CertificateOptions(verify=False)
+        self._config = config
+
+        self._options_novalidate = CertificateOptions()
+
+        # Check if we're using a custom list of a CA certificates
+        if isinstance(config.federation_custom_ca_list, OpenSSLCertificateAuthorities):
+            self._options_validate = CertificateOptions(
+                # Use custom CA trusted root certs
+                trustRoot=config.federation_custom_ca_list,
+            )
+            return
+
+        # If not, verify using those provided by the operating environment
+        self._options_validate = CertificateOptions(
+            # Use CA root certs provided by OpenSSL
+            trustRoot=platformTrust(),
+        )
 
     def get_options(self, host):
         # Use _makeContext so that we get a fresh OpenSSL CTX each time.
-        return ClientTLSOptions(host, self._options._makeContext())
+
+        # Check if certificate verification has been enabled
+        if (self._config.federation_verify_certificates and
+                host not in self._config.federation_certificate_validation_whitelist):
+            # Require verification
+            return ClientTLSOptions(host, self._options_validate._makeContext())
```

Comment:
Note that `ClientTLSOptions` is a copy of the real one in twisted, which skips a verification step, so you need to change somethign here